libcache/cache.c

```diff
@@ -0,0 +1,667 @@
+			CLEAR_DIRTY(linePtr->flags);
+		}
+	}
+
+	return position;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j, err;
+	uint64_t addr;
+	ssize_t writeCount;
+	cacheline_t *linePtr;
+
+	for (i = 0; i < cache->numSets; ++i) {
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			linePtr = &(cache->sets[i].lines[j]);
+
+			if (!IS_VALID(linePtr->flags)) {
+				continue;
+			}
+
+			if (IS_DIRTY(linePtr->flags)) {
+				addr = cache_computeAddr(cache, linePtr->tag, i);
+				writeCount = cache_flushLine(cache, linePtr, addr);
+
+				if (writeCount < (ssize_t)cache->lineSize) {
+					return -EIO;
+				}
+			}
+
+			free(linePtr->data);
+			CLEAR_VALID(linePtr->flags);
+		}
+	}
+
+	err = resourceDestroy(cache->lock);
+	if (err < 0) {
+		return err;
+	}
+
+	free(cache->sets);
+	free(cache);
+
+	return EOK;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_executePolicy(cachectx_t *cache, cacheline_t *linePtr, uint64_t addr, int policy)
+{
+	ssize_t writeCount = cache->lineSize;
+
+	if (policy == 1) {
+		writeCount = cache_flushLine(cache, linePtr, addr);
+	}
+
+	return writeCount;
+}
+
+
+static cacheline_t *cache_allocateLine(cachectx_t *cache, const uint64_t setIndex, const uint64_t tag)
+{
+	int i, lineIndex;
+	uint64_t addr = 0;
+	cacheline_t *temp = NULL, *linePtr = NULL;
+	unsigned char flags = 0;
+	void *data;
+	ssize_t writeCount;
+	cacheset_t *setPtr = &cache->sets[setIndex];
+
+	if (setPtr->count < LIBCACHE_NUM_WAYS) {
+		setPtr->count++;
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (!IS_VALID(setPtr->lines[i].flags)) {
+				lineIndex = i;
+				break;
+			}
+		}
+
+		temp = &(setPtr->lines[lineIndex]);
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (setPtr->tags[i] == NULL) {
+				setPtr->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&setPtr->timestamps, temp);
+	}
+	else {
+		temp = setPtr->timestamps;
+
+		if (IS_DIRTY(temp->flags)) {
+			addr = cache_computeAddr(cache, temp->tag, setIndex);
+			writeCount = cache_flushLine(cache, temp, addr);
+
+			if (writeCount < (ssize_t)cache->lineSize) {
+				return NULL;
+			}
+		}
+
+		LIST_REMOVE(&setPtr->timestamps, temp);
+		free(temp->data);
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (setPtr->lines[i].tag == temp->tag) {
+				lineIndex = i;
+				break;
+			}
+		}
+
+		LIST_ADD(&setPtr->timestamps, temp);
+	}
+
+	linePtr = &(setPtr->lines[lineIndex]);
+
+	data = malloc(cache->lineSize * sizeof(unsigned char));
```

Comment:
sizeof(unsigned char) = 1 by definition
```suggestion
	data = malloc(cache->lineSize);
```